feat(shim): support archive layers

vdice opened the pull request:
Closes https://github.com/spinkube/containerd-shim-spin/issues/28

Depends on https://github.com/fermyon/spin/pull/2523

Draft for now as 2523 is based on Spin's main branch which has since bumped wasmtime to 21.0.1.  Can either wait for that same bump to occur in this project (presumably in tandem with pinning to the next Spin release with said version) or can explore backporting 2523 to the v2.5 branch and cutting a patch release with it.

TODO
- [x] Depends on PR(s) bumping spin + co repos to next release tag and wasmtime to 21.0.1 in Cargo.toml.  Rebase after this lands.
  - [x] https://github.com/spinkube/containerd-shim-spin/pull/144
- [x] decide on merge approach (avoid wasmtime/spin crates bump or wait for them to land)
  - I think waiting for the upcoming 2.6 Spin release and bumping wasmtime to suit will be fine; no big hurry to get this in.
- [x] add test app/test case

jsturtevant commented on containerd-shim-spin/src/engine.rs:
```diff
@@ -140,6 +136,30 @@ impl SpinEngine {
+                                log::debug!(
+                                    "<<< TMPDIR is not set; setting to current working directory for unpacking archive layer"
+                                );
+                                env::set_var("TMPDIR", ".");
```
I believe this is because we are using a scratch empty dir for configuring the containers root dir that the shim is running in so there isn't anything in it at all.

This generally makes sense as a solution.  Is there any concern with using `.` which is relative?  I don't think there is but might be better to be protective and use `env::current_dir()`?

vdice replied:
Thanks @jsturtevant, that definitely seems better.  I used an `unwrap_or(".")` but maybe we should just hard error if env::current_dir() errors?

jsturtevant replied:
`unwrap_or(".")` works for me 👍

vdice commented:
In https://github.com/spinkube/containerd-shim-spin/pull/121/commits/3f3d2919604b7424545b24d1bf9dbb6bf352ff94 I added a new `spin-static-assets` app and incorporated into the integration tests, to cover the archive layer functionality.  I haven't had success running the integration tests locally, so leaning on CI here to check.  I'm also happy to relegate this commit (or general approach if adjustments are requested) as a follow-up instead.

Mossaka commented on containerd-shim-spin/src/engine.rs:
```diff
@@ -40,10 +40,6 @@ const SPIN_HTTP_LISTEN_ADDR_ENV: &str = "SPIN_HTTP_LISTEN_ADDR";
 const RUNTIME_CONFIG_PATH: &str = "/runtime-config.toml";
 /// Describes an OCI layer with Wasm content
 const OCI_LAYER_MEDIA_TYPE_WASM: &str = "application/vnd.wasm.content.layer.v1+wasm";
```
suggestion: make `spin_oci::client::WASM_LAYER_MEDIA_TYPE` public and use it in the Spin Shim.

vdice replied:
In fact, Spin itself is waiting on the canonical type/string upstream: https://github.com/fermyon/spin/blob/main/crates/oci/src/client.rs#L34-L35.  To be defined in https://github.com/bytecodealliance/registry perhaps?  Therefore expecting both Spin and the shim would pull this from the upstream crate.

Mossaka reviewed (approved):
LGTM!

kate-goldenring commented on containerd-shim-spin/src/engine.rs:
```diff
@@ -367,6 +376,25 @@ impl SpinEngine {
+        // which will fall back to /tmp if TMPDIR is not set. /tmp is either not found or not accessible
+        // in the shim environment, hence setting to current working directory.
+        if env::var("TMPDIR").is_err() {
+            log::debug!(
```
When do we expect `TMPDIR` to be set? I am wondering if this log is necessary.

vdice replied:
Good point... it appears it isn't ever set, currently.  Glad to remove this debug log if you'd like.

jsturtevant replied:
I think we would want to leave it, otherwise we might overwrite it. I don't know what scenario it would be set in, but if it was set and we overwrite I don't think that would be an expected behavior?

kate-goldenring reviewed (approved):
LGTM!

jsturtevant reviewed (approved):
LGTM

vdice commented:
Thanks, all.  Please merge when ready.
